src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -369,22 +373,36 @@ public Task CheckForAddedFileBeingOpenMaybeAsync(bool useAsync, ImmutableArray<s
                             // and if it was actually open, we'll schedule an update asynchronously.
                             if (TryOpeningDocumentsForFilePathCore(w, newFileName, textBuffer, hierarchy: null))
                             {
-                                // The files are now tied to the buffer, but let's schedule work to correctly update the context.
-                                var token = _asynchronousOperationListener.BeginAsyncOperation(nameof(CheckForAddedFileBeingOpenMaybeAsync));
-                                UpdateContextAfterOpenAsync(newFileName).CompletesAsyncOperation(token);
+                                // We've opened the document, but we only have to go to the UI thread if either:
+                                //
+                                // 1. We have multiple documents with the same file path, so we need to ensure the context is actually correct.
+                                // 2. We haven't enabled the suggested actions source provider, which needs to be done once.
+                                var needsContextUpdate = w.CurrentSolution.GetDocumentIdsWithFilePath(newFileName).Length >= 2;
+                                if (needsContextUpdate || !_suggestedActionSourceProviderEnabled)
```

Comment:
>_suggestedActionSourceProviderEnabled [](http://example.com/codeflow?start=59&length=37)

Is it ok for this to change from false=>true during the loop?

I'm still curious whether it would make sense to only call UpdateContextAfterOpenAndEnableSuggestedActionsSourceProviderAsync once instead of inside the loop.

Reply:
> I'm still curious whether it would make sense to only call UpdateContextAfterOpenAndEnableSuggestedActionsSourceProviderAsync once instead of inside the loop.

That would reduce the potential number of UI switches. The upper bound of switches though here are files that are open and loaded prior to the solution loading, which isn't going to be that many. This fix is also being applied simultaneously to 17.14 and main so I'm keeping the change as small as reasonably possible since there's a desire for quick servicing.

Reply:
> Is it ok for this to change from false=>true during the loop?

Yep, the only concern is the UI thread gets that call at least once. Beyond that, doesn't matter.